### ISSUE TITLE
feat: Support for OracleDB thick mode

### ISSUE THREAD

**Issue by kangshung** [ibis-project/ibis#6913]
### Is your feature request related to a problem?

I consider using a tool that uses ibis as a dependency. I need to connect to Oracle DB. I am aware that only THICK python-oracledb clients are permitted to connect to my DB.

I see that [ibis uses the python-oracledb "thin client" to connect to Oracle databases](https://ibis-project.org/backends/oracle/#connecting-to-older-oracle-databases), which means that if it's true, I won't be able to use it.

Is my thinking correct?

### Describe the solution you'd like

I would like ibis to support oracledb thick mode: https://python-oracledb.readthedocs.io/en/latest/user_guide/initialization.html#enabling-python-oracledb-thick-mode

### What version of ibis are you running?

6.1.0

### What backend(s) are you using, if any?

Oracle

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by gforsyth** [comment]
Hi @kangshung -- thanks for raising this.

I don't know if we've confirmed whether or not thick mode will work -- do you want to try it out?  I don't have the requisite library files to enable thick-mode, which is one reason we haven't yet tried it.

My understanding (perhaps wrong) is that if you first do

```python
import oracledb

oracledb.init_oracle_client()
```

then you can try to connect using `ibis` as you would with the thin-client, but now `oracledb` will be a thick-client instead.

**Comment by kangshung** [comment]
Hi @gforsyth 

I tried to connect, and good news is that using the Oracle client seemed to be working fine.

However, ibis appears to be using the `database` parameter as a `sid` rather than the `service_name` (https://python-oracledb.readthedocs.io/en/latest/api_manual/connect_param.html#connectparams-methods). Additionally, when I don't wish to set any database, ibis inserts SID = FREE into the dsn. This makes me impossible to connect with my database.

Also, it looks like ibis does not support a `dsn` parameter at all, which could solve the issue (in pure oracledb it just works) (https://python-oracledb.readthedocs.io/en/latest/user_guide/connection_handling.html#net-service-names-for-connection-strings). If `dsn` was added as a parameter that overwrites host, port, and service_name, as far as I know, it would be possible to connect with all possible combinations.

**Comment by gforsyth** [comment]
Hey @kangshung -- thanks for trying it out!

I'm about to push up a draft PR that allows specifying either `sid`, `service_name`, or passing in a `dsn` string and hopefully that should unblock you.

**Comment by gforsyth** [comment]
For my reference (and others) -- here is a table listing the feature differences between thin mode and thick mode in `python-oracledb`: https://python-oracledb.readthedocs.io/en/latest/user_guide/appendix_a.html

**Comment by gforsyth** [comment]
There should be full `dsn` support on `master` now -- but ping here and we can reopen if issues persist (or open a new issue, whichever is easier).

**Comment by kangshung** [comment]
Thank you very much for the update. Looks like connection looks fine, but further problem arrived. #7164 #7165
